### PR TITLE
Schema export as a "minimalistic" JSON.

### PR DESCRIPTION
**Pull request by dkorolev**

### CODE REVIEW

File: TypeSystem/Schema/test.cc [hunk 1]
```diff
@@ -213,13 +217,21 @@ TEST(Schema, SmokeTestFullStruct) {
             restored_schema.Describe<Language::FSharp>());
   EXPECT_EQ(FileSystem::ReadFileAsString(Golden("smoke_test_struct.md")),
             restored_schema.Describe<Language::Markdown>());
+
+  // Don't just `EXPECT_EQ(golden, ReadFileAsString("golden/...))`, but compare re-generated JSON,
+  // as the JSON file in the golden directory is pretty-printed.
+
```
Comment (mzhurovich) [inline TypeSystem/Schema/test.cc]: Extra new line?

File: TypeSystem/Serialization/test.cc [hunk 2]
```diff
@@ -115,6 +116,35 @@ CURRENT_STRUCT(WithTime) {
...
+struct B_VariantName {
+  static const char* VariantNameImpl() { return "B"; }
+};
+using B = NamedVariant<B_VariantName, B_VariantTypes>;
+
+using E_VariantTypes = TypeListImpl<A, B>;
+struct E_VariantName {
+  static const char* VariantNameImpl() { return "E"; }
+};
+using E = NamedVariant<E_VariantName, E_VariantTypes>;
```
Comment (mzhurovich) [inline TypeSystem/Serialization/test.cc]: Using `E` for `Variant` in this type test is a bit confusing - I was thinking about some `Enum` first :)

File: TypeSystem/types.h [hunk 3]
```diff
@@ -44,12 +44,28 @@ struct CurrentSuper {
 // The superclass for all Current structures.
 struct CurrentStruct : CurrentSuper {};
 
-#define IS_CURRENT_STRUCT(T) (std::is_base_of<::current::CurrentStruct, ::current::decay<T>>::value)
-
 // The superclass for `Variant` type.
 struct CurrentVariant : CurrentSuper {};
 
+template<typename NAME>
```
Comment (mzhurovich) [inline TypeSystem/types.h]: Run `make indent` for the whole dir?

File: TypeSystem/types.h [hunk 4]
```diff
@@ -44,12 +44,28 @@ struct CurrentSuper {
...
+  static const char* VariantNameImpl() {
+    return "Variant";
+  }
+};
+
+#define IS_CURRENT_STRUCT(T) (std::is_base_of<::current::CurrentStruct, ::current::decay<T>>::value)
 #define IS_VARIANT(T) (std::is_base_of<::current::CurrentVariant, ::current::decay<T>>::value)
+#define IS_CURRENT_STRUCT_OR_VARIANT(T) (std::is_base_of<::current::CurrentSuper, ::current::decay<T>>::value)
+
+#define IS_NONEMPTY_CURRENT_STRUCT(T) \
```
Comment (mzhurovich) [inline TypeSystem/types.h]: Did you decide to declare `IS_NONEMPTY`for some usability reasons? At first glance, `IS_EMPTY` is more obvious choice :)

File: TypeSystem/variant.h [hunk 5]
```diff
@@ -135,13 +135,16 @@ struct VariantTypeCheckedAssignment<TypeListImpl<TS...>> {
...
 struct VariantImpl;
 
-template <typename... TYPES>
-struct VariantImpl<TypeListImpl<TYPES...>> : CurrentVariant {
+template <typename NAME, typename... TYPES>
+struct VariantImpl<NAME, TypeListImpl<TYPES...>> : CurrentVariantImpl<NAME> {
   using typelist_t = TypeListImpl<TYPES...>;
   using DEPRECATED_T_(TYPELIST) = typelist_t;
+
+  using variant_t = VariantImpl<NAME, typelist_t>;
```
Comment (mzhurovich) [inline TypeSystem/variant.h]: s/`variant_t`/`variant_impl_t`/?
`Variant<>::variant_t` looks strange.

File: TypeSystem/variant.h [hunk 6]
```diff
@@ -304,21 +307,25 @@ struct VariantImpl<TypeListImpl<TYPES...>> : CurrentVariant {
 };
 
 // `Variant<...>` can accept either a list of types, or a `TypeList<...>`.
-template <typename T, typename... TS>
+template <typename NAME, typename T, typename... TS>
```
Comment (mzhurovich) [inline TypeSystem/variant.h]: Can't we just use `typename... TS` instead of `typename T, typename... TS` here?

File: TypeSystem/Schema/schema.h [hunk 7]
```diff
@@ -496,12 +512,188 @@ struct LanguageSyntaxImpl<Language::Markdown> {
...
+  // LCOV_EXCL_START
+  static std::string ErrorMessageWithTypeId(TypeID type_id, FullSchemaPrinter&) {
+    return "#error \"Unknown struct with `type_id` = " + current::ToString(type_id) + "\"\n";
+  }
+  // LCOV_EXCL_STOP
+};
+
+template <>
+struct LanguageSyntaxImpl<Language::JSON> final {
+  // The `Language::JSON` format ignores `Header()/Footer()`, all work is is done by `struct FullSchemaPrinter`.
```
Comment (mzhurovich) [inline TypeSystem/Schema/schema.h]: s/`is is`/`is`/

File: TypeSystem/Schema/test.cc (hunk 1 above)
Comment (dkorolev) [inline TypeSystem/Schema/test.cc]: :+1: 7f3ef7e

File: TypeSystem/Serialization/test.cc (hunk 2 above)
Comment (dkorolev) [inline TypeSystem/Serialization/test.cc]: Yes, it was short for `Event`, with which I was testing. Renamed into `Q`. :-)

File: TypeSystem/types.h (hunk 3 above)
Comment (dkorolev) [inline TypeSystem/types.h]: Done.

File: TypeSystem/variant.h (hunk 5 above)
Comment (dkorolev) [inline TypeSystem/variant.h]: :+1: d7dac38080b3ebaf35db4315383bdb3a879d7051

File: TypeSystem/variant.h (hunk 6 above)
Comment (dkorolev) [inline TypeSystem/variant.h]: This is to make sure no empty `Variant<>` can be declared.

File: TypeSystem/Schema/schema.h (hunk 7 above)
Comment (dkorolev) [inline TypeSystem/Schema/schema.h]: :+1: c4db3e3

File: TypeSystem/types.h (hunk 4 above)
Comment (dkorolev) [inline TypeSystem/types.h]: The reason was indeed convoluted. Look, this:
```
if (IS_CURRENT_STRUCT(X) && !IS_NONEMPTY_CURRENT_STRUCT(X)) {
```
is now this:
```
if (IS_EMPTY_CURRENT_STRUCT(X)) {
```

Two good news are: we have tests, and they passed. :+1: bbfc94a

**Comment by dkorolev** [comment]
Hi @mzhurovich 

Wow, this pull request was created three days ago. Feels like two weeks have passed. Certainly enough lines of code to review at ones. Sorry again, hope individual commits make sense.

Should be enough implemented by now to use with F#. This integration feels cursed, which makes me even more eager to push it through.

Thanks,
Dima

File: TypeSystem/Serialization/test.cc [hunk 8]
```diff
@@ -742,6 +777,86 @@ TEST(Serialization, VariantAsJSON) {
...
 
+TEST(Serialization, NamedVariantAsJSON) {
+  using namespace serialization_test::named_variant;
+
+  {
+    Q e;
+    A a;
+    X x;
+    a = x;
+    e = a;
```
Comment (mzhurovich) [inline TypeSystem/Serialization/test.cc]: This choice of type and variable names is really mind-blowing :)

File: TypeSystem/typename.h [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+// For JSON serialization and other repetitive operations, enable returning type name as a `const char*`.
+template <typename T, typename>
+struct CurrentTypeNameAsConstCharPtrImpl {
+  static const char* Value() {
+    struct LocalSingleton {
+      std::string s;
+    };
+    auto& singleton = ThreadLocalSingleton<LocalSingleton>();
+    singleton.s = CurrentTypeName<T>();
+    return singleton.s.c_str();
```
Comment (mzhurovich) [inline TypeSystem/typename.h]: To be honest, I'm confused with this black magic of mixing `const char*`/`std::string` return types :-/

File: TypeSystem/Schema/golden/smoke_test_struct.cc [hunk 10]
```diff
@@ -34,20 +34,23 @@ struct Empty {
...
-  int32_t x;
+  E e;
 };
+using MyFreakingVariant = Variant<A, X, Y>;
 struct C {
   Empty e;
-  Variant<X, Y> c;
+  MyFreakingVariant c;
 };
+using Variant_B_A_B_C_Empty_E = Variant<A, B, C, Empty>;
```
Comment (mzhurovich) [inline TypeSystem/Schema/golden/smoke_test_struct.cc]: `B_A_B_C_Empty_E` means something? ;)

File: TypeSystem/Schema/schema.h [hunk 11]
```diff
@@ -251,44 +273,51 @@ struct LanguageSyntaxCPP : CurrentStructPrinter<CPP_LANGUAGE_SELECTOR> {
     // `operator()`-s of this block print the complete declaration of a type in C++, if necessary.
     // Effectively, they ignore everything but `struct`-s.
```
Comment (mzhurovich) [inline TypeSystem/Schema/schema.h]: Update comment?

File: TypeSystem/Schema/golden/smoke_test_struct.cc (hunk 10 above)
Comment (mzhurovich) [inline TypeSystem/Schema/golden/smoke_test_struct.cc]: Ok, found out that you use prefix `B_` and suffix `_E`. Trying to find out, why :)

**Comment by mzhurovich** [comment]
@dkorolev Such clean code, much wow :)

File: TypeSystem/Serialization/test.cc (hunk 8 above)
Comment (dkorolev) [inline TypeSystem/Serialization/test.cc]: :+1: 2c46391

File: TypeSystem/typename.h (hunk 9 above)
Comment (dkorolev) [inline TypeSystem/typename.h]: It's for:
a) repetitive operations, such as serialize an array of 1'000 variants into JSON, to not create the description each time, and
b) for RapidJSON to not have to take ownership of an `std::string` containing the type name.
I just used a thread-local singleton to cache `std::string`-s for the duration of the thread using `CurrentTypeNameAsConstCharPtr<T>` for a particular `T`.

File: TypeSystem/Schema/golden/smoke_test_struct.cc (hunk 10 above)
Comment (dkorolev) [inline TypeSystem/Schema/golden/smoke_test_struct.cc]: It's analogous to C / C++ name mangling by the linker. Purpose: for name of `Variant<Variant<Foo, Bar>, Baz>` to make sense.

File: TypeSystem/Schema/schema.h (hunk 11 above)
Comment (dkorolev) [inline TypeSystem/Schema/schema.h]: :+1: 2c46391

**Comment by dkorolev** [comment]
@mzhurovich Thanks! PTAL :)

File: TypeSystem/typename.h (hunk 9 above)
Comment (mzhurovich) [inline TypeSystem/typename.h]: How is it optimized if `singleton.s = CurrentTypeName<T>();` is executed in each call to `Value()`?

Comment (dkorolev) [inline TypeSystem/typename.h]: Oh, this one. It must have to do with me being an idiot. Let me un-D'Bill-ize it real quick.

Comment (dkorolev) [inline TypeSystem/typename.h]: :+1: b3e0ed1

**Pull request merged**